server/blob/blob-memory/src/main/java/org/apache/james/blob/memory/MemoryBlobStore.java

```diff
@@ -70,16 +76,35 @@ public MemoryBlobStore(BlobId.Factory factory) {
 
     @Override
     public Mono<byte[]> readBytes(BucketName bucketName, BlobId blobId) {
-        return Mono.fromCallable(() -> retrieveStoredValue(blobId));
+        Preconditions.checkNotNull(bucketName);
+        return Mono.fromCallable(() -> retrieveStoredValue(bucketName, blobId));
     }
 
     @Override
     public InputStream read(BucketName bucketName, BlobId blobId) {
-        return new ByteArrayInputStream(retrieveStoredValue(blobId));
+        Preconditions.checkNotNull(bucketName);
+        return new ByteArrayInputStream(retrieveStoredValue(bucketName, blobId));
     }
 
-    private byte[] retrieveStoredValue(BlobId blobId) {
-        return Optional.ofNullable(blobs.get(blobId))
-            .orElseThrow(() -> new ObjectStoreException("unable to find blob with id " + blobId));
+    @Override
+    public Mono<Void> deleteBucket(BucketName bucketName) {
+        Preconditions.checkNotNull(bucketName);
+
+        synchronized (blobs) {
```

Comment:
You synchronize mono creation, not mono execution. /!\ things will not behave like you think in a concurrent environment...

You should write it like this:

```
Mono<Void> deleteBucket(blah...) {
   // preconditions blah

   return Mono.fromRunnable(() -> {
        synchronized (blobs) {
            // delete logic here
        }
   });
}